File: solution/py/maximumSubarraySum.py
class Solution(object):
    # Faster prevIdx Version
    def maximumSubarraySum(self, nums, k):
        res = 0
        prevIdx = {}  # num -> previous index
        curSum = 0
        l = 0

        for r in range(len(nums)):
            curSum += nums[r]

            i = prevIdx.get(nums[r], -1)

            while l <= i or r - l + 1 > k:
                curSum -= nums[l]
                l += 1

            if r - l + 1 == k:
                res = max(res, curSum)

            prevIdx[nums[r]] = r
        return res
    

    # freq count map version
    def maximumSubarraySum(nums, k):
    res = 0
    count = {}
    curSum = 0
    l = 0

    for r in range(len(nums)):
        curSum += nums[r]
        count[nums[r]] = count.get(nums[r], 0) + 1

        if r - l + 1 > k:
            count[nums[l]] -= 1
            if count[nums[l]] == 0:
                del count[nums[l]]
            curSum -= nums[l]
            l += 1

        if len(count) == r - l + 1 and r - l + 1 == k:
            res = max(res, curSum)

    return res

    

s = Solution()

print(s.maximumSubarraySum([2, 2, 3, 4, 5], 3))

File: solution/java/maximumSubarraySum.java
import java.util.*;

public class maximumSubarraySum {
    public static void main(String[] args) {
        System.out.println(maximumSubarraySum(new int[] { 2, 2, 3, 4, 5 }, 3));
    }
    // prevIndex faster version
    public static int maximumSubarraySum(int[] nums, int k) {
        int res = 0;
        Map<Integer, Integer> prevIdx = new HashMap<>();
        int curSum = 0;
        int l = 0;

        for (int r = 0; r < nums.length; r++) {
            curSum += nums[r];
            int i = prevIdx.getOrDefault(nums[r], -1);

            while (l <= i || r - l + 1 > k) {
                curSum -= nums[l];
                l++;
            }

            if (r - l + 1 == k) {
                res = Math.max(res, curSum);
            }

            prevIdx.put(nums[r], r);
        }

        return res;
    }


    //count (frequency map) Version
    public int maximumSubarraySum(int[] nums, int k) {
        int res = 0;
        Map<Integer, Integer> count = new HashMap<>();
        int curSum = 0;
        int l = 0;

        for (int r = 0; r < nums.length; r++) {
            curSum += nums[r];
            count.put(nums[r], count.getOrDefault(nums[r], 0) + 1);

            if (r - l + 1 > k) {
                count.put(nums[l], count.get(nums[l]) - 1);
                if (count.get(nums[l]) == 0) {
                    count.remove(nums[l]);
                }
                curSum -= nums[l];
                l++;
            }

            if (count.size() == r - l + 1 && r - l + 1 == k) {
                res = Math.max(res, curSum);
            }
        }

        return res;
    
}

}
